src/graph.vala

```diff
@@ -0,0 +1,307 @@
+     * cache.
+     */
+
+    public int vincount {get; internal set;}
+    public int voutcount {get; internal set;}
+    static int _counter = 0;
+
+    construct
+    {
+        vincount = 0;
+        voutcount = 0;
+    }
+
+    public Variable.dummy()
+    {
+        var name = "D%08d".printf(_counter);
+        base(name : name);
+        _counter++;
+    }
+
+    public Variable()
+    {
+        var name1 = "V%08d".printf(_counter);
+        base(name : name1);
+        _counter++;
+    }
+    public string to_string()
+    {
+        var type = "unknown";
+        if(vincount == voutcount) {
+            type = "dummy";
+        }
+        if(vincount == 0) {
+            type = "output";
+        }
+        if(voutcount == 0) {
+            type = "output";
+        }
+        return "%s:%s(as vin %d, as vout %d)".printf(
+                name, type, vincount, voutcount);
+    }
+}
+
+public abstract class Vast.Network.Operation : Object
+{
+    public int nin {get; construct set;}
+    public int nout {get; construct set;}
+    public string name {get; construct set;}
+
+    /* This is the symbol that we use to look up the UFunc via GIR */
+    public Operation (string name,
+        int nin,
+        int nout)
+    {
+        base(name : name, nin : nin, nout : nout);
+    }
+
+    public abstract void
+    prepare(Object [] inputs,
+            Object? [] outputs);
+        /* For any operation and given inputs, we shall be able to infer the output types*/
+
+        /* check if the input arrays are correct */
+
+        /* create array in outputs if the item is null */
+
+    public abstract void
+    execute(Object[] inputs, Object[] outputs);
+        /* carry out the operation */
+
+        /* Look up the operation via GIR */
+
+        /* call it. */
+}
+
+public class Vast.Network.Node: Object
+{
+    public Vast.Network.Operation operation;
+    public Vast.Network.Variable [] vout;
+    public Vast.Network.Variable [] vin;
+
+    public Node(
+        /* FIXME: can we use Object(a : a) type constructor here? */
+        Vast.Network.Operation operation,
+        Vast.Network.Variable [] vin,
+        Vast.Network.Variable [] vout
+        )
+    {
+        this.operation = operation;
+        this.vin = vin;
+        this.vout = vout;
+    }
+
+    public string
+    to_string()
+    {
+        var sb = new StringBuilder();
+        foreach(var v in vout) {
+            sb.append_printf("%s, ", v.name);
+        }
+        sb.append_printf("= %s (", operation.name);
+        foreach(var v in vin) {
+            sb.append_printf("%s, ", v.name);
+        }
+        sb.append_printf(")");
+        return sb.str;
+    }
+}
+
+public class Vast.Network.Graph : Object
+{
+    public List<Vast.Network.Node> nodes;
+    public List<Vast.Network.Variable> variables;
+
+    /* find the node that generates V */
+    public Vast.Network.Node?
+    find_node(Vast.Network.Variable V)
+    {
+        /** FIXME: exception or null? **/
```

Comment:
Always prefer returning `null` in these situations (e.g. finding an element in a data structure). Not having a node in the graph does not constitute an error condition and it's perfectly possible that no node generates the given variable in this case.